### PR TITLE
feat(Auth) support preferPrivateSession flag

### PR DESCRIPTION
**Pull request by SuTran**
*Issue #880*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_flutter/lib/categories/amplify_auth_category.dart [hunk 1]
```diff
@@ -207,8 +207,10 @@ class AuthCategory {
         : throw _pluginNotAddedException("Auth");
   }
 
-  Future<SignInResult> signInWithWebUI({AuthProvider? provider}) {
-    var request = SignInWithWebUIRequest(provider: provider);
+  Future<SignInResult> signInWithWebUI(
```
Comment (fjnoyp) [inline packages/amplify_flutter/lib/categories/amplify_auth_category.dart:210]: Could you update the category API to match this signature instead: 

Future<SignInResult> signInWithWebUI({AuthProvider? provider, SignInWithWebUIOptions options})

Then within amplify_auth_plugin_interface you could create the SignInWithWebUIOptions class. 

Then within amplify_auth_cognito, you could create CognitoSignInWIthWebUIOptions which would have the field "isPreferPrivateSession". 

We discussed internally about this best long term approach for this change, and believe that having this options object would give us the greatest long term flexibility while also avoiding cognito specific and platform specific implementations from leaking into the auth category file.

**Review by fjnoyp — CHANGES_REQUESTED** [review]
Hi @SuTran thanks a ton for getting this PR out so fast!  We really appreciate your help with implementation here. 

I left a comment about the API changes.  We want to use an options parameter instead of directly passing the boolean.  More explanations in the comments.

File: packages/amplify_flutter/lib/categories/amplify_auth_category.dart (hunk 1 above)
Comment (SuTran) [reply packages/amplify_flutter/lib/categories/amplify_auth_category.dart:210]: Thanks @fjnoyp for your feedback.

I will modify the `amplify_auth_category.dart` to have `SignInWithWebUIOptions` options. It makes sense to me too.

I saw within `amplify_auth_plugin_interface.dart` there is a class called `SignInWithWebUIRequest`. If I created `SignInWithWebUIOptions`, so what is the key different between "Request" and "Options", I think design this way will lead to confusion later when we adding more params to support new features.

Can you share more thoughts on this?

Comment (fjnoyp) [reply packages/amplify_flutter/lib/categories/amplify_auth_category.dart:210]: Hi.  Thanks!

The key difference is that "Request" is internal to our library and not exposed in the public API.  While "Options" would be a method parameter in the public API, `amplify_auth_category`

Comment (SuTran) [reply packages/amplify_flutter/lib/categories/amplify_auth_category.dart:210]: Thanks. I have updated the code. Can you help to check

**Review by fjnoyp — APPROVED** [review]
Looks great @SuTran thanks again for submitting this PR. 

I'll contact some others on my team to get you more approvals for a merge.

File: packages/amplify_auth_cognito/ios/Classes/FlutterSignInWithWebUIRequest.swift [hunk 2]
```diff
@@ -20,9 +20,16 @@ import amplify_core
...
+  var isPreferPrivateSession: Bool = false
 
   init(dict: NSMutableDictionary){
     self.provider = getAuthProvider(provider: dict["authProvider"] as! String?)
+    var options = dict["options"] as! Dictionary<String, Any>?
+    if (options != nil) {
+      if (options!["isPreferPrivateSession"] != nil && options!["isPreferPrivateSession"] as! Bool) {
+        self.isPreferPrivateSession = options!["isPreferPrivateSession"] as! Bool
+      }
+    }
```
Comment (dnys1) [inline packages/amplify_auth_cognito/ios/Classes/FlutterSignInWithWebUIRequest.swift:32]:
```suggestion
    if let options = dict["options"] as? [String: Any] {
       self.isPreferPrivateSession = options["isPreferPrivateSession"] as! Bool
    }
```

File: packages/amplify_auth_cognito/lib/src/CognitoSignIn/CognitoSignInWithWebUIOptions.dart [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+class CognitoSignInWithWebUIOptions extends SignInWithWebUIOptions {
+  bool? isPreferPrivateSession;
+  CognitoSignInWithWebUIOptions({this.isPreferPrivateSession}) : super();
+  Map<String, dynamic> serializeAsMap() {
+    final Map<String, dynamic> pendingRequest = {
+      'isPreferPrivateSession': isPreferPrivateSession
+    };
+    pendingRequest.removeWhere((_, v) => v == null);
+    return pendingRequest;
+  }
```
Comment (dnys1) [inline packages/amplify_auth_cognito/lib/src/CognitoSignIn/CognitoSignInWithWebUIOptions.dart:27]:
```suggestion
  @override
  Map<String, dynamic> serializeAsMap() => <String, dynamic>{
        'isPreferPrivateSession': isPreferPrivateSession,
      };
```

File: packages/amplify_auth_plugin_interface/lib/src/SignIn/SignInWithWebUIOptions.dart [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+abstract class SignInWithWebUIOptions {
+  const SignInWithWebUIOptions();
+  Map<String, dynamic> serializeAsMap() {
+    throw UnimplementedError(
+        'serializeAsMap() has not been implemented on SignInWithWebUIOptions.');
+  }
```
Comment (dnys1) [inline packages/amplify_auth_plugin_interface/lib/src/SignIn/SignInWithWebUIOptions.dart:21]: Prefer abstract method

```suggestion
  Map<String, dynamic> serializeAsMap();
```

File: packages/amplify_auth_cognito/lib/src/CognitoSignIn/CognitoSignInWithWebUIOptions.dart [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import 'package:amplify_auth_plugin_interface/amplify_auth_plugin_interface.dart';
+
+class CognitoSignInWithWebUIOptions extends SignInWithWebUIOptions {
+  bool? isPreferPrivateSession;
```
Comment (dnys1) [inline packages/amplify_auth_cognito/lib/src/CognitoSignIn/CognitoSignInWithWebUIOptions.dart:19]: Copied from iOS code:

```suggestion
  /// iOS-only: Starts the webUI signin in a private browser session, if supported by the current browser.
  ///
  /// Note that this value internally sets `prefersEphemeralWebBrowserSession` in ASWebAuthenticationSession.
  /// As per Apple documentation, Whether the request is honored depends on the user’s default web browser.
  /// Safari always honors the request.
  /// 
  /// Defaults to `false`.
  final bool isPreferPrivateSession;
```

File: packages/amplify_auth_cognito/lib/src/CognitoSignIn/CognitoSignInWithWebUIOptions.dart [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  CognitoSignInWithWebUIOptions({this.isPreferPrivateSession}) : super();
```
Comment (dnys1) [inline packages/amplify_auth_cognito/lib/src/CognitoSignIn/CognitoSignInWithWebUIOptions.dart:20]: Prefer default value here

```suggestion
  const CognitoSignInWithWebUIOptions({this.isPreferPrivateSession = false});
```

**Review by dnys1 — COMMENTED** [review]
Looks good, thanks for the PR!

Needs some minor style things and doc comments, but otherwise it's good to go 👍

File: packages/amplify_auth_cognito/lib/src/CognitoSignIn/CognitoSignInWithWebUIOptions.dart (hunk 3 above)
Comment (SuTran) [reply packages/amplify_auth_cognito/lib/src/CognitoSignIn/CognitoSignInWithWebUIOptions.dart]: Thank you!. I have updated the code

File: packages/amplify_auth_plugin_interface/lib/src/SignIn/SignInWithWebUIOptions.dart (hunk 4 above)
Comment (SuTran) [reply packages/amplify_auth_plugin_interface/lib/src/SignIn/SignInWithWebUIOptions.dart]: Thank you! I have updated the code

File: packages/amplify_auth_cognito/ios/Classes/FlutterSignInWithWebUIRequest.swift (hunk 2 above)
Comment (SuTran) [reply packages/amplify_auth_cognito/ios/Classes/FlutterSignInWithWebUIRequest.swift]: Thank you! I have updated the code

File: packages/amplify_auth_cognito/lib/src/CognitoSignIn/CognitoSignInWithWebUIOptions.dart (hunk 5 above)
Comment (SuTran) [reply packages/amplify_auth_cognito/lib/src/CognitoSignIn/CognitoSignInWithWebUIOptions.dart]: Thank you! I have updated the code

File: packages/amplify_auth_cognito/lib/src/CognitoSignIn/CognitoSignInWithWebUIOptions.dart (hunk 6 above)
Comment (SuTran) [reply packages/amplify_auth_cognito/lib/src/CognitoSignIn/CognitoSignInWithWebUIOptions.dart]: Thank you! I have updated the code

File: packages/amplify_auth_cognito/ios/Classes/FlutterSignInWithWebUIRequest.swift [hunk 7]
```diff
@@ -20,9 +20,14 @@ import amplify_core
 
 struct FlutterSignInWithWebUIRequest {
   var provider: AuthProvider?
+  var isPreferPrivateSession: Bool = false
 
   init(dict: NSMutableDictionary){
     self.provider = getAuthProvider(provider: dict["authProvider"] as! String?)
+    var options = dict["options"] as! Dictionary<String, Any>?
```
Comment (dnys1) [inline packages/amplify_auth_cognito/ios/Classes/FlutterSignInWithWebUIRequest.swift:27]: Please remove this line

**Review by dnys1 — APPROVED** [review]
Looks great! Minor change and it's good to go!

Thank you again 😄

**Pull request merged**